apps/examples/iperf2/compat/delay.c

```diff
@@ -0,0 +1,326 @@
+#  ifdef HAVE_NANOSLEEP
+    // Don't call nanosleep for values less than 10 microseconds
+    // as the syscall is too expensive.  Let the busy loop
+    // provide the delay for times under that.
+    if (nsec_adjusted > 10000) {
+	nanosleep(&requested, &remaining);
+    }
+#  endif
+    while (1) {
+	clock_gettime(CLOCK_REALTIME, &t2);
+	if (timespec_greaterthan(t2, finishtime))
+	    break;
+    }
+    // Compute the delay error in units of nanoseconds
+    // and cast to type double
+    err = (double) (timespec_diff(t2, t1) - (usec * 1000));
+    // printf("req: %ld adj: %f err: %.5f (ns)\n", usec, nsec_adjusted, kalmanerr.x);
+    kalman_update(&kalmanerr, err);
+}
+#endif // HAVE_KALMAN
+#else
+// Sadly, these systems must use the not so efficient gettimeofday()
+// and working units are microseconds, struct is timeval
+static void timeval_add_ulong (struct timeval *tv0, unsigned long value) {
+    tv0->tv_usec += value;
+    if (tv0->tv_usec >= MILLION) {
+	tv0->tv_sec++;
+	tv0->tv_usec -= MILLION;
+    }
+}
+static inline
+int timeval_greaterthan(struct timeval tv1, struct timeval tv0) {
+    if (tv1.tv_sec > tv0.tv_sec ||					\
+	((tv0.tv_sec == tv1.tv_sec) && (tv1.tv_usec > tv0.tv_usec))) {
+	return 1;
+    } else {
+	return 0;
+    }
+}
+// tv1 assumed greater than tv0
+static double timeval_diff (struct timeval tv1, struct timeval tv0) {
+    double result;
+    if (tv1.tv_usec < tv0.tv_usec) {
+	tv1.tv_usec += MILLION;
+	tv1.tv_sec--;
+    }
+    result = (double) (((tv1.tv_sec - tv0.tv_sec) * MILLION) + (tv1.tv_usec - tv0.tv_usec));
+    return result;
+}
+void delay_busyloop (unsigned long usec) {
+    struct timeval t1, t2;
+    gettimeofday( &t1, NULL );
+    timeval_add_ulong(&t1, usec);
+    while (1) {
+	gettimeofday( &t2, NULL );
+	if (timeval_greaterthan(t2, t1))
+	    break;
+    }
+}
+#ifdef HAVE_KALMAN
+// Request units is microseconds
+// Adjust units is microseconds
+void delay_kalman (unsigned long usec) {
+    struct timeval t1, t2, finishtime;
+    long usec_adjusted;
+    double err;
+    static kalman_state kalmanerr={
+	0.00001, //q process noise covariance
+	0.1, //r measurement noise covariance
+	0.0, //x value, error predictio (units nanoseconds)
+	1, //p estimation error covariance
+	0.25 //k kalman gain
+    };
+    // Get the current clock
+    gettimeofday( &t1, NULL );
+    // Perform the kalman adjust per the predicted delay error
+    if (kalmanerr.x > 0) {
+	usec_adjusted = usec - (long) floor(kalmanerr.x);
+	if (usec_adjusted < 0)
+	    usec_adjusted = 0;
+    }
+    else
+	usec_adjusted = usec + (long) floor(kalmanerr.x);
+    // Set the finishtime
+    finishtime = t1;
+    timeval_add_ulong(&finishtime, usec_adjusted);
+#  ifdef HAVE_NANOSLEEP
+    // Don't call nanosleep for values less than 10 microseconds
+    // as the syscall is too expensive.  Let the busy loop
+    // provide the delay for times under that.
+    if (usec_adjusted > 10) {
+	struct timespec requested={0,0}, remaining;
```

Comment:
same as above.

Reply:
it's opensource so It would be better not to apply coding rule